### ISSUE TITLE
Tdef doesn't restart after error

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#4658]
## Steps to reproduce

```supercollider
Tdef(\x, { loop { 0.1.wait; Error("some error").throw } }).play; // ok.
Tdef(\x, { loop { 0.1.wait; Error("some error").throw } }); // ok.
Tdef(\x, { loop { 0.1.wait; Error("some error").throw } }); // not ok.
```

## Expected vs. actual behavior

The Tdef should recover each time.

Maybe related: #4647

**Comment by telephon** [comment]
This was introduced in https://github.com/supercollider/supercollider/commit/27dbc319309126003bea45d89184a665dc46d09d

**Comment by jamshark70** [comment]
Why is the Tdef object placed into the Pprotect, instead of the Prout?

I think the disposable object -- the one actually encountering the error -- should be the protected object, not the persistent wrapper.

**Comment by telephon** [comment]
I'm not sure I understand what you mean. 

maybe something like 
`this.asStream.asProtected`
instead of
`this.asProtected.asStream`
?

As things currently are, the player (PauseStream) must explicitly be informed of the error, and this is not possible by default because the stream itself knows nothing about the player.

**Comment by jamshark70** [comment]
TL;DR I think this is a better solution: 1. Remove `asProtected`. 2. Apply Pprotect to the source pattern, not to the Tdef's overall stream.

```
	source_ { arg obj;
		pattern = Pprotect(
			if(obj.isKindOf(Function)) { this.convertFunction(obj) } { obj },
			{ this.player.streamError }
		);
		if (obj.isNil) { pattern = this.class.default; source = obj; };
		this.wakeUp;
		source = obj;
		this.changed(\source, obj);
	}
```

Rationale:

If you use `protect` in a routine, expected behavior is that the exception handler is cleared afterward.

```
(
r = fork {
	protect {
		thisThread.exceptionHandler.debug("exception handler in protect");
		nil.xyz;
	};
};
)

exception handler in protect: a Function

// afterward...
r.exceptionHandler
-> nil
```

So, Tdef's current assumption that its stream's exception handler will still be in place after an error contradicts the general behavior of `protect` (and Pprotect should be consistent with this general behavior).

If you reset a stream that includes a `protect`, then the stream starts from the beginning and re-applies `protect`.

```
r.reset.play;
exception handler in protect: a Function
```

So Tdef should be doing something more like that.

In general, I think the best approach to situations like this is that the wrapper (Tdef) is persistent and the resources that it uses (players, streams) may be created and disposed of as needed. I think this is the cause of the current problem: Pprotect (and, really, most places in the class library) assume that an error means that the stream is dead, but here, the thing being "protected" is not a stream, but rather the persistent wrapper. So the attempt to recover does not re-invoke Pprotect -- which then leads to the (IMO mistaken) suggestion that Pprotect's effect should be persistent as well.

I had thought at first that the best approach would be to re-create the player when there is an error. Unfortunately, PauseStream has some bugs about its state variables and methods, and I couldn't get it to pick up the previous playing/not-playing state correctly:

```
// both of the following are WTF

// if I don't refresh, then it "was stopped"
// how could it have been stopped in the past?
// It's brand-new (it never played)
// so this prevents wakeUp from re-playing
p = PauseStream(Routine { 1.yield });
p.wasStopped
-> true

// but if I do refresh, then it "is playing" (but it never played)
// and this also prevents wakeUp from re-playing
p = PauseStream(Routine { 1.yield }).refresh;
p.isPlaying
-> true
```

So I had to abandon that approach. (But this approach does work for me in my BP class, where the behavior on error is to stop, allow the user to correct the error, and then the user manually re-plays.)

If Pprotect is added to the source, then changing the source re-invokes Pprotect and a new exception handler is in place.

This also handles `Tdef(\x).reset.play` after error -- because the source includes Pprotect, restarting the source from the beginning also puts a new exception handler in place.

BTW `Pprotect` has another bug, dating back to the original implementation:

```
(
r = Pprotect(
	Pprotect(
		Prout { nil.xyz },
		{ "inner exception handler".postln }
	),
	{ "outer exception handler".postln }
).play;
)

-> an EventStreamPlayer
inner exception handler

// outer exception handler never executes... :-|
```

**Comment by jamshark70** [comment]
While at it: There's a formatting issue in https://github.com/supercollider/supercollider/blob/87779bdcbc42279c8e6cc28291115bd46cc59209/SCClassLibrary/JITLib/Patterns/Pdef.sc#L367-L368 -- method closing brace not on its own line.

**Comment by jamshark70** [comment]
The nested Pprotect error may be fixed this way:

```
Pprotect : FilterPattern {
	var <>func;
	*new { arg pattern, func;
		^super.new(pattern).func_(func)
	}
	storeArgs { ^[ pattern, func ] }
	asStream {
		var rout = Routine(pattern.embedInStream(_)),
		next = thisThread.exceptionHandler;
		rout.exceptionHandler = { |error|
			// 'func' might throw an error
			// we must clear the exceptionHandler before that
			// otherwise, infinite recursion is the result
			rout.exceptionHandler = nil;
			func.value(error, rout);
			next.handleError(error)
		};
		^rout
	}
}
```

**Comment by jamshark70** [comment]
Actually... forget all that. Sorry for noise, but I had a better idea.

The real problem is that Pprotect's desired state (with exception handler) is not encapsulated into its stream. It's *applied to* the stream initially, but not under control of the stream itself.

We expect that resetting a stream will restart everything from the beginning. For Pprotect to do that, it should assign the exception handler *within* the stream returned by `asStream`, not externally.

So the following fixes both the nested Pprotect bug *and* the Tdef bug (without any changes to Tdef).

```
Pprotect : FilterPattern {
	var <>func;
	*new { arg pattern, func;
		^super.new(pattern).func_(func)
	}
	storeArgs { ^[ pattern, func ] }
	asStream {
		var next = thisThread.exceptionHandler;
		^Routine({ |inval|
			thisThread.exceptionHandler = { |error|
				// 'func' might throw an error
				// we must clear the exceptionHandler before that
				// otherwise, infinite recursion is the result
				thisThread.exceptionHandler = nil;
				func.value(error, thisThread);
				next.handleError(error)
			};
			pattern.embedInStream(inval)
		})
	}
}
```

BUT there is one case about this that makes me uneasy. I can't find a solution right now. The above will work as long as asStream is called only when needed. If asStream is called in advance, and then used in another location that has an exception handler of its own, then the other exception handler won't be used.

```
r = Pprotect(Prout { nil.xyz }, { "r error handler".postln }).asStream;
q = Pprotect(r, { "q error handler" }).asStream;
q.next;

r error handler  // no "q error handler"
ERROR: Message 'xyz' not understood.
```

I'm out of time at the moment. I think this last issue is a rare case but would be worth some attention.

**Comment by telephon** [comment]
Yes, nested protects should be possible. There is still something not completely right.

Here is a version that doesn't need a separate Routine, and follows more closely the implementation of `prTry`.  It passes all tests. I think this is a first step, but still won't pass the ball up to the outer stream after it has handled the error:

```supercollider
Pprotect : FilterPattern {
	var <>func;
	*new { arg pattern, func;
		^super.new(pattern).func_(func)
	}
	storeArgs { ^[ pattern, func ] }

	embedInStream { arg inval;
		// keep previous handler
		var next = thisThread.exceptionHandler;
		thisThread.exceptionHandler = { |error|
			// 'func' might throw an error
			// we must clear the exceptionHandler before that
			// otherwise, infinite recursion is the result
			thisThread.exceptionHandler = nil;
			func.value(error, thisThread);
			thisThread.exceptionHandler = next;
			next.handleError(error)
		};
		inval = pattern.embedInStream(inval);
		// restore previous handler
		thisThread.exceptionHandler = next;
		^inval
	}
}
```

**Comment by telephon** [comment]
This is how the error should percolate up the parent stream series:

```supercollider

(
var r = Routine {
	var r = Routine {
		var r = Routine {
			Error("test").throw;
		};
		r.exceptionHandler = { "bing".postln };
		r.next;
	};
	r.exceptionHandler = { "bang".postln };
	r.next;
};
r.exceptionHandler = { "bong".postln };
r.next;
)


(
var r = Routine {
	var r = Routine {
		var r = Routine {
			Error("test").throw;
		};
		var next = thisThread;
		r.exceptionHandler = { "bing".postln; next.handleError };
		r.next;
	};
	var next = thisThread;
	r.exceptionHandler = { "bang".postln; next.handleError };
	r.next;
};
var next = thisThread;
r.exceptionHandler = { "bong".postln; next.handleError };
r.next;
)
```

**Comment by jamshark70** [comment]
Yes, it's a better approach to put this in embedInStream.

I was just trying to replicate `protect` logic within Pprotect:embedInStream, when I realized something quite unfortunate: `Pprotect` is semantically incompatible with `protect`.

`protect` is used for a cleanup that should always fire, whether or not there is an error: `file = File(...); protect { ... do stuff with file ... } { file.close }`. If there is an error, you want the file to be closed. If there was no error, you still want the file to be closed.

`Pprotect` assumes that error recovery (in the typical cases I'm aware of, informing a player that something has gone wrong) should occur *only* in case of error. If there was no error, then, no cleanup.

`Pseq([Pprotect(..., { player.status = \stopped }), Pprotect(..., ...)])`

... but the player will not actually have finished at this point.

So it's neither Pprotect nor Ptry -- it's a hybrid that doesn't exist anywhere else in the class library.

This should be explained carefully in the documentation, then.

Anyway, this resolves the "upward throw" problem, without changing Pprotect's semantics (`result.throw` is not the approach suggested in the previous post, but it's the technique used in `Function:protect`, so... good enough for me):

```
Pprotect : FilterPattern {
	var <>func;
	*new { arg pattern, func;
		^super.new(pattern).func_(func)
	}
	storeArgs { ^[ pattern, func ] }

	embedInStream { arg inval;
		var result = prTry { pattern.embedInStream(inval) };
		if(result.isException) {
			func.value(result, thisThread);
			result.throw;  // to nested error handlers
		} {
			// here, `protect` calls the error handler
			// but Pprotect has a slightly different meaning
			^result
		};
	}
}
```

Also, the uneasy case that I mentioned turns out to be a general problem, not specific to patterns. So we could a/ go ahead with this proposed form of Pprotect, and b/ file another issue for this `protect` bug (which, tbh, I have no idea how to solve).

```
r = Routine { protect { nil.xyz } { "r error handler".postln } };
q = Routine { protect { r.embedInStream } { "q error handler" } };
q.next;

// only "r error handler" fires
```

**Comment by telephon** [comment]
> q = Routine { protect { r.embedInStream } { "q error handler" } };

ahhg, yes, I also overlooked this.

The solution above is much better, thank you!

**Comment by telephon** [comment]
P.S. several times I had thought about the name `Pprotect`, and indeed it is a little misleading. I couldn't find a better one, so I kept it.

In the future, we may not need the use case in TaskProxy anymore, if `streamError` is called by the throw itself. But that is something for a separate discussion.

**Comment by telephon** [comment]
fixed by  #4659
